### ISSUE TITLE
Implement Java plugin

### ISSUE THREAD

**Issue by DmitrySoshnikov** [DmitrySoshnikov/syntax#14]
We need to support LR-parser generator for Java. See [example plugin](https://github.com/DmitrySoshnikov/syntax/tree/master/src/plugins/example), or [C# plugin](https://github.com/DmitrySoshnikov/syntax/tree/master/src/plugins/csharp) which is closer to Java.

**Comment by enebo** [comment]
I am very interested in a Java plugin as JRuby uses an ancient LALR parser called Jay.  It would be nice to switch to something both being improved and used by more than 2 projects on the planet :)

I also have a killer grammar file to convert and see how things work out: https://github.com/jruby/jruby/blob/master/core/src/main/java/org/jruby/parser/RubyParser.y ... Ruby is a pretty massive grammar.

**Comment by DmitrySoshnikov** [comment]
@enebo, yeah, we should add it, it was a long standing issue. Great you have the grammar example! I'll take a look at this plugin soon (don't have the latest best practicing in Java, so would appreciate the review).

**Comment by enebo** [comment]
@DmitrySoshnikov I have also been wanting to get more into Rust so I am keen on trying this out.

**Comment by DmitrySoshnikov** [comment]
As of `v0.1.10` Java plugin is supported. See the [README.md](https://github.com/DmitrySoshnikov/syntax/blob/master/src/plugins/java/README.md) for details.

Try it out, and let me know if you see any issues!

**Comment by DmitrySoshnikov** [comment]
For now it uses the same states parsing as in JavaScript, see [the example](https://github.com/DmitrySoshnikov/syntax/blob/master/src/plugins/java/templates/lr.template.java#L239-L240), doing a lot of `Integer.valueOf(...)` conversions such as `"s5" -> 5`.

Alternatively we can explore using another states format, to avoid string parsing, and use e.g. enum. E.g. instead of `"s5"` (which is "Shift-5"), to use `new TableEntry(TableEntryType.SHIFT, 5)`. Not sure yet which is faster/better for Java.

**Comment by enebo** [comment]
@DmitrySoshnikov Thanks.  I will try things out.
